## Import das bibliotecas

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.neural_network import MLPClassifier

## Conexão com o Google Drive e import do dataset

In [2]:
from google.colab import drive

# conexão do colab com o google drive
drive.mount('/content/drive')

# criando um datafram com a primeira aba do dataset
sonar = pd.read_excel('/content/drive/My Drive/sonar.xlsx', sheet_name=0)

Mounted at /content/drive


In [3]:
sonar.head()

Atributo_1_62  Atributo_2_63  ...  Atributo_60_121  Classe
0         0.0200         0.0371  ...           0.0032   Rocha
1         0.0453         0.0523  ...           0.0044   Rocha
2         0.0262         0.0582  ...           0.0078   Rocha
3         0.0100         0.0171  ...           0.0117   Rocha
4         0.0762         0.0666  ...           0.0094   Rocha

[5 rows x 61 columns]

## Tratamento no dataset para aplicar os modelos

In [4]:
# passando todas as linhas na variavel X usando iloc para ler todas as informações
X = sonar.iloc[:, 0:(sonar.shape[1] - 1)]

# aplicando label enconder para converter variaveis categoricas em númericas e treinando o modelo em todas as features(colunas)
le = LabelEncoder()
y = le.fit_transform(sonar.iloc[:, (sonar.shape[1] - 1)])

# gravando num objetvo os nomes das classes
class_names = le.classes_

In [5]:
# dividindo as bases em bases de treino e test com 10% para test e 90% para treino
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0, test_size = 0.1)

## Decision tree

In [6]:
# chamando o metodo decisiontreeclassifier
decision_tree = tree.DecisionTreeClassifier()

# treinando a base de treino
decision_tree = decision_tree.fit(X_train, y_train)

# prevendo a base de test
y_pred = decision_tree.predict(X_test)

# imprime a acurácia da base de treino
print('Classificador Árvore de decisão:\n')
print('Acurácia da base de treinamento: {:.2f}'.format(decision_tree.score(X_train, y_train)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + "(prev)" for x in class_names])
print(conf_table)

Classificador Árvore de decisão:

Acurácia da base de treinamento: 1.00
              precision    recall  f1-score   support

        Mina       0.58      0.88      0.70         8
       Rocha       0.89      0.62      0.73        13

    accuracy                           0.71        21
   macro avg       0.74      0.75      0.71        21
weighted avg       0.77      0.71      0.72        21

       Mina(prev)  Rocha(prev)
Mina            7            1
Rocha           5            8


## Random Forest

In [ ]:
# chamando o metodo Random Forest
random_forest = RandomForestClassifier(n_estimators=10)

# treinando a base de treino
random_forest = random_forest.fit(X_train, y_train)

# prevendo a base de test
y_pred = random_forest.predict(X_test)

# imprime a acurácia da base de treino
print('Classificador Random Forest:\n Number of estimators = 10\n')
print('Acurácia da base de treinamento: {:.2f}'.format(random_forest.score(X_train, y_train)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + "(prev)" for x in class_names])
print(conf_table)

Classificador Random Forest:
 Number of estimators = 10

Acurácia da base de treinamento: 0.99
              precision    recall  f1-score   support

        Mina       0.60      0.75      0.67         8
       Rocha       0.82      0.69      0.75        13

    accuracy                           0.71        21
   macro avg       0.71      0.72      0.71        21
weighted avg       0.74      0.71      0.72        21

       Mina(prev)  Rocha(prev)
Mina            6            2
Rocha           4            9


## Extra tree

In [ ]:
# chamando o metodo Extra tree
extra_tree = ExtraTreesClassifier(n_estimators=10)

# treinando a base de treino
extra_tree = extra_tree.fit(X_train, y_train)

# prevendo a base de test
y_pred = extra_tree.predict(X_test)

# imprime a acurácia da base de treino
print('Classificador Extreme Tree:\n Number of estimators = 10\n')
print('Acurácia da base de treinamento: {:.2f}'.format(extra_tree.score(X_train, y_train)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + "(prev)" for x in class_names])
print(conf_table)

Classificador Extreme Tree:
 Number of estimators = 10

Acurácia da base de treinamento: 1.00
              precision    recall  f1-score   support

        Mina       0.58      0.88      0.70         8
       Rocha       0.89      0.62      0.73        13

    accuracy                           0.71        21
   macro avg       0.74      0.75      0.71        21
weighted avg       0.77      0.71      0.72        21

       Mina(prev)  Rocha(prev)
Mina            7            1
Rocha           5            8


## Adaboost

In [ ]:
# chamando o metodo Adaboost
ada_boost = AdaBoostClassifier(tree.DecisionTreeClassifier(max_depth=1),
                               learning_rate=0.1, algorithm="SAMME", n_estimators=5)

# treinando a base de treino
ada_boost.fit(X_train, y_train)

# prevendo a base de test
y_pred = ada_boost.predict(X_test)

# imprime a acurácia da base de treino
print('Classificador AdaBoost:\n Number os estimators = 5\n')
print('Acurácia da base de treinamento: {:.2f}'.format(ada_boost.score(X_train, y_train)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + "(prev)" for x in class_names])
print(conf_table)

Classificador AdaBoost:
 Number os estimators = 5

Acurácia da base de treinamento: 0.78
              precision    recall  f1-score   support

        Mina       0.54      0.88      0.67         8
       Rocha       0.88      0.54      0.67        13

    accuracy                           0.67        21
   macro avg       0.71      0.71      0.67        21
weighted avg       0.75      0.67      0.67        21

       Mina(prev)  Rocha(prev)
Mina            7            1
Rocha           6            7


In [ ]:
# chamando o metodo Adaboost
ada_boost = AdaBoostClassifier(tree.DecisionTreeClassifier(max_depth=1),
                                learning_rate=0.01, n_estimators=5)

# treinando a base de treino
ada_boost.fit(X_train, y_train)

# prevendo a base de test
y_pred = ada_boost.predict(X_test)

# imprime a acurácia da base de treino
print('Classificador AdaBoost:\n Number os estimators = 5\n')
print('Acurácia da base de treinamento: {:.2f}'.format(ada_boost.score(X_train, y_train)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + "(prev)" for x in class_names])
print(conf_table)

Classificador AdaBoost:
 Number os estimators = 5

Acurácia da base de treinamento: 0.77
              precision    recall  f1-score   support

        Mina       0.54      0.88      0.67         8
       Rocha       0.88      0.54      0.67        13

    accuracy                           0.67        21
   macro avg       0.71      0.71      0.67        21
weighted avg       0.75      0.67      0.67        21

       Mina(prev)  Rocha(prev)
Mina            7            1
Rocha           6            7


## Redes neurais com uma camada oculta

In [ ]:
# chamando o metodo Redes neurais
mlp = MLPClassifier(solver='lbfgs', random_state=0, hidden_layer_sizes=[100], learning_rate_init=0.01, max_iter=200000)

# treinando a base de treino
mlp.fit(X_train, y_train)

# prevendo a base de test
y_pred = mlp.predict(X_test)

# imprimindo os parâmetros da rede neural
print('Camadas da rede: {}'.format(mlp.n_layers_))
print('Neurônios na hidden layer: {}'.format(mlp.hidden_layer_sizes))
print('Neurônios na layer de saída: {}'.format(mlp.n_outputs_))
print('Weights na layer de entrada: {}'.format(mlp.coefs_[0].shape))
print('Weights na hidden layer: {}'.format(mlp.coefs_[1].shape))

# imprime a acurácia da base de treino
print('Acurácia da base de treinamento: {:.2f}'.format(mlp.score(X_train, y_train)))

# imprime a acurácia da base de teste
print('Acurácia da base de teste: {:.2f}'.format(mlp.score(X_test, y_test)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + "(prev)" for x in class_names])
print(conf_table)

Camadas da rede: 3
Neurônios na hidden layer: [100]
Neurônios na layer de saída: 1
Weights na layer de entrada: (60, 100)
Weights na hidden layer: (100, 1)
Acurácia da base de treinamento: 1.00
Acurácia da base de teste: 0.86
              precision    recall  f1-score   support

        Mina       0.78      0.88      0.82         8
       Rocha       0.92      0.85      0.88        13

    accuracy                           0.86        21
   macro avg       0.85      0.86      0.85        21
weighted avg       0.86      0.86      0.86        21

       Mina(prev)  Rocha(prev)
Mina            7            1
Rocha           2           11


## Redes neurais com duas camada oculta

In [ ]:
# chamando o metodo Redes neurais
mlp = MLPClassifier(solver='lbfgs', random_state=0, hidden_layer_sizes=[100, 60], learning_rate_init=0.01, max_iter=200000)

# treinando a base de treino
mlp.fit(X_train, y_train)

# prevendo a base de test
y_pred = mlp.predict(X_test)

# imprimindo os parâmetros da rede neural
print('Camadas da rede: {}'.format(mlp.n_layers_))
print('Neurônios na camada oculta: {}'.format(mlp.hidden_layer_sizes))
print('Neurônios na camada de saída: {}'.format(mlp.n_outputs_))
print('Pesos na camada de entrada: {}'.format(mlp.coefs_[0].shape))
print('Pesos na camada oculta: {}'.format(mlp.coefs_[1].shape))

# imprime a acurácia da base de treino
print('Acurácia da base de treinamento: {:.2f}'.format(mlp.score(X_train, y_train)))

# imprime a acurácia da base de teste
print('Acurácia da base de teste: {:.2f}'.format(mlp.score(X_test, y_test)))

# imprime o classification report da previsão e do teste
print(classification_report(y_test, y_pred, target_names=class_names))

# aplica matrix de confusão na previsão e no teste
conf_matrix = confusion_matrix(y_test, y_pred)

# criar a tabela em datafram da matrix de confusão, carregando os nomes das features
conf_table = pd.DataFrame(data=conf_matrix, index=class_names, columns=[x + '(prev)' for x in class_names])
print(conf_table)

Camadas da rede: 4
Neurônios na camada oculta: [100, 60]
Neurônios na camada de saída: 1
Pesos na camada de entrada: (60, 100)
Pesos na camada oculta: (100, 60)
Acurácia da base de treinamento: 1.00
Acurácia da base de teste: 0.86
              precision    recall  f1-score   support

        Mina       0.86      0.75      0.80         8
       Rocha       0.86      0.92      0.89        13

    accuracy                           0.86        21
   macro avg       0.86      0.84      0.84        21
weighted avg       0.86      0.86      0.86        21

       Mina(prev)  Rocha(prev)
Mina            6            2
Rocha           1           12


## Conclusão

1. Utilizei os modelos de: árvore de decisão, extratree, adaboost um padrão e 
outro alterando os parâmetros de prunning e redes neurais.
Nas redes neurais utilizei topologias do modelo com uma camada oculta e com duas camadas ocultas, variei alguns parâmetros além do número de camadas, como o número de neurônios na hidden layer, melhorou o precision e recall da previsão de rocha.

2. Apenas as acurácias de treinamento dos modelos Adaboost não foram 99% ou 100%, tivemos acurácias de 77% e 83% no modelos do Adaboost, pode ser um indicativo de Underfitting, pois esperava-se uma acurácia maior no dataset de treinamento
Em relação ao precision, tivemos valores elevados para Rocha em média 90% e valores baixos para Mina e média 70%, com destaque para o modelo de redes neurais que teve precision mais elevado para Mina 78%.
Sobre recall tivemos valores elevados para Mina em média 90% e valores baixos para Rocha, em média 70%, com destaque novamente para o modelo de redes neurais com recall de 88%

3. Fiz uns testes alterando entre os modelos de algoritmo de prunning do Adaboost, quando alterei do SAMME para o SAMME.R e inseri no segunda learning_rate=0.01, tive uma ligeira queda na acurácia, precision e recall do modelo. Contudo quando tiro learning_rate os resultados ficam identicos, ou seja sem alterações nos resultados quando alteramos de algoritmo. Quando insiro learning_rate temos uma queda na acurácia, precision e recall, tendo maior acurácia no learning rate 0.1 se comparado a 0.01